src/hotspot/share/utilities/intrusiveList.hpp

```diff
@@ -0,0 +1,1508 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_UTILITIES_INTRUSIVELIST_HPP
+#define SHARE_UTILITIES_INTRUSIVELIST_HPP
+
+#include "memory/allStatic.hpp"
+#include "metaprogramming/enableIf.hpp"
+#include "metaprogramming/logical.hpp"
+#include "utilities/align.hpp"
+#include "utilities/debug.hpp"
+#include "utilities/macros.hpp"
+#include <type_traits>
+
+class IntrusiveListEntry;
+class IntrusiveListImpl;
+
+/**
+ * The IntrusiveList class template provides a doubly-linked list in
+ * which the links between elements are embedded directly into objects
+ * contained in the list.  As a result, there are no copies involved
+ * when inserting objects into the list or referencing list objects,
+ * and removing an object from a list need not involve destroying the
+ * object.
+ *
+ * To be used in a IntrusiveList, an object must have a
+ * IntrusiveListEntry member.  An IntrusiveList is associated with the
+ * class of its elements and the entry member.
+ *
+ * An object can be in multiple lists at the same time, so long as
+ * each list uses a different entry member.  That is, the class of the
+ * object must have multiple IntrusiveListEntry members, one for each
+ * list the object is simultaneously an element.
+ *
+ * The API for IntrusiveList is modelled on the containers provided by
+ * the C++ standard library.  In particular, iteration over the
+ * elements is provided via iterator classes.
+ *
+ * IntrusiveLists support polymorphic elements.  Because the objects
+ * in a list are externally managed, rather than being embedded values
+ * in the list, the actual type of such objects may be more specific
+ * than the list's element type.
+ *
+ * * T is the class of the elements in the list.  Must be a possibly
+ * const-qualified class type.
+ *
+ * * entry_member is a pointer to class member referring to the
+ * IntrusiveListEntry subobject of T used by this list.
```

Comment:
> nit, "subobject" confused me a little. Maybe "element" or "member" ?

"subobject" (C++14 1.8/2) is the correct term.  "member" would have worked,
but doesn't with the change to use a function accessor rather than a
pointer-to-data-member.